PR: https://github.com/odoo/odoo/pull/173288

From: 75b385bae036031b8fef651977bad71691b1e7a2
From: Yannick Tivisse
Date: 2024-08-26 10:55:02

Breaking data model changes scores: del:4.8000000000000025 + add:4, change matches:
-    applicant_id = fields.Many2one('hr.applicant', string="Applicant", index='btree_not_null', ondelete='set null')
-    name = fields.Char("Subject / Application", required=True, help="Email subject for applications sent via email", index='trigram')
-    email_from = fields.Char("Email", size=128, compute='_compute_partner_phone_email',
+    email_from = fields.Char(related="candidate_id.email_from", readonly=False)
+    categ_ids = fields.Many2many('hr.applicant.category', string="Tags", compute='_compute_categ_ids', store=True, readonly=False)
-    partner_phone = fields.Char("Phone", size=32, compute='_compute_partner_phone_email',
-    partner_phone_sanitized = fields.Char(string='Sanitized Phone Number', compute='_compute_partner_phone_sanitized', store=True, index='btree_not_null')
-    partner_mobile = fields.Char("Mobile", size=32, compute='_compute_partner_phone_email',
-    partner_mobile_sanitized = fields.Char(string='Sanitized Mobile Number', compute='_compute_partner_mobile_sanitized', store=True, index='btree_not_null')
-    emp_id = fields.Many2one('hr.employee', string="Employee", help="Employee linked to the applicant.", copy=False)
+    applicant_ids = fields.One2many('hr.applicant', 'candidate_id')
+    email_from = fields.Char(
+    partner_phone = fields.Char(
+    partner_phone_sanitized = fields.Char(
+    partner_mobile = fields.Char(
+    partner_mobile_sanitized = fields.Char(
+    meeting_ids = fields.One2many('calendar.event', 'candidate_id', 'Meetings')
+    attachment_ids = fields.One2many('ir.attachment', 'res_id', domain=[('res_model', '=', 'hr.candidate')], string='Attachments')
-    applicant_id = fields.One2many('hr.applicant', 'emp_id', 'Applicant', groups="hr.group_hr_user")
+    candidate_id = fields.One2many('hr.candidate', 'employee_id', 'Candidate', groups="hr.group_hr_user")

Total Changes: 844

[IMP] hr_recruitment: Introduce hr.candidate model

Purpose
=======

Based on this task https://www.odoo.com/odoo/9285/tasks/4037988/9285/project.task/4050307?debug=1&cids=1
and many feedbacks we got before (leading partially to Headhunter industry), we realize that if we
want to cover the whole market of ATS we need a strong solution.

Specification
=============

We split hr.applicant between candidates and applications

A candidate is created outside of a job position, it contains all data that will
never change (skills, name, contact info, expected salary), it gathers history
(applied job positions, states, ...)

An application is the link between an candidate and a Job Position, it contains all related
data (source, medium, proposed salary, evaluation, ...) and it got a state
(first interview, contract proposed, refused)

In a Job position, you would see distinct applicants, but through the hr.applications model.
This mechanism is similar to a vehicle which is the link between a Model and a Driver.

The list view "all applications" would list "All Applicants" and each Applicant would have a
smartbutton "Applications"

closes odoo/odoo#173288

Details: https://link.excalidraw.com/l/65VNwvy7c4X/5SjjwEHMIPG
Taskid: 4050442
Related: odoo/enterprise#66734
Related: odoo/upgrade#6300
Signed-off-by: Yannick Tivisse (yti) <yti@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_recruitment/models/__init__.py
+++ b/addons/hr_recruitment/models/__init__.py
@@ -1,7 +1,10 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
 from . import hr_department
 from . import hr_applicant
 from . import hr_applicant_category
 from . import hr_applicant_refuse_reason
+from . import hr_candidate
 from . import hr_recruitment_degree
 from . import hr_recruitment_source
 from . import hr_recruitment_stage
@@ -12,6 +15,7 @@ from . import calendar
 from . import digest
 from . import utm_campaign
 from . import utm_source
+from . import res_company
 from . import res_users
 from . import ir_ui_menu
 from . import mail_activity_plan

--- a/addons/hr_recruitment/models/calendar.py
+++ b/addons/hr_recruitment/models/calendar.py
@@ -6,57 +6,83 @@ from odoo.exceptions import AccessError
 
 
 class CalendarEvent(models.Model):
-    """ Model for Calendar Event """
     _inherit = 'calendar.event'
 
-    @api.model_create_multi
-    def create(self, vals_list):
-        events = super(CalendarEvent, self).create(vals_list)
-        try:
-            self.env['hr.applicant'].check_access_rights('read')
-        except AccessError:
-            return events
-
-        if "default_applicant_id" in self.env.context:
-            applicant_attachments = self.env['hr.applicant'].browse(self.env.context['default_applicant_id']).attachment_ids
-            for event in events:
-                self.env['ir.attachment'].create([{
-                    'name': att.name,
-                    'type': 'binary',
-                    'datas': att.datas,
-                    'res_model': event._name,
-                    'res_id': event.id
-                } for att in applicant_attachments])
-        return events
-
     @api.model
     def default_get(self, fields):
+        self_ctx = self
         if self.env.context.get('default_applicant_id'):
-            self = self.with_context(
-                default_res_model='hr.applicant', #res_model seems to be lost without this
+            self_ctx = self.with_context(
+                default_res_model='hr.applicant',  # res_model seems to be lost without this
                 default_res_model_id=self.env.ref('hr_recruitment.model_hr_applicant').id,
                 default_res_id=self.env.context.get('default_applicant_id'),
                 default_partner_ids=self.env.context.get('default_partner_ids'),
                 default_name=self.env.context.get('default_name')
             )
+        elif self.env.context.get('default_candidate_id'):
+            self_ctx = self.with_context(
+                default_res_model='hr.candidate',  # res_model seems to be lost without this
+                default_res_model_id=self.env.ref('hr_recruitment.model_hr_candidate').id,
+                default_res_id=self.env.context.get('default_candidate_id'),
+                default_partner_ids=self.env.context.get('default_partner_ids'),
+                default_name=self.env.context.get('default_name')
+            )
 
-        defaults = super(CalendarEvent, self).default_get(fields)
+        defaults = super(CalendarEvent, self_ctx).default_get(fields)
 
         # sync res_model / res_id to opportunity id (aka creating meeting from lead chatter)
         if 'applicant_id' not in defaults:
-            res_model = defaults.get('res_model', False) or self.env.context.get('default_res_model')
-            res_model_id = defaults.get('res_model_id', False) or self.env.context.get('default_res_model_id')
-            if (res_model and res_model == 'hr.applicant') or (res_model_id and self.env['ir.model'].sudo().browse(res_model_id).model == 'hr.applicant'):
-                defaults['applicant_id'] = defaults.get('res_id', False) or self.env.context.get('default_res_id', False)
+            res_model = defaults.get('res_model', False) or self_ctx.env.context.get('default_res_model')
+            res_model_id = defaults.get('res_model_id', False) or self_ctx.env.context.get('default_res_model_id')
+            if (res_model and res_model == 'hr.applicant') or (res_model_id and self_ctx.env['ir.model'].sudo().browse(res_model_id).model == 'hr.applicant'):
+                defaults['applicant_id'] = defaults.get('res_id', False) or self_ctx.env.context.get('default_res_id', False)
 
         return defaults
 
+    applicant_id = fields.Many2one('hr.applicant', string="Applicant", index='btree_not_null', ondelete='set null')
+    candidate_id = fields.Many2one(
+        'hr.candidate',
+        string="Candidate",
+        compute="_compute_candidate_id",
+        store=True,
+        readonly=False,
+        index='btree_not_null',
+        ondelete='set null')
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        events = super().create(vals_list)
+        try:
+            self.env['hr.applicant'].check_access_rights('read')
+        except AccessError:
+            return events
+
+        attachments = False
+        if "default_applicant_id" in self.env.context:
+            attachments = self.env['hr.applicant'].browse(self.env.context['default_applicant_id']).attachment_ids
+        elif "default_candidate_id" in self.env.context:
+            attachments = self.env['hr.candidate'].browse(self.env.context['default_candidate_id']).attachment_ids
+        if attachments:
+            self.env['ir.attachment'].create([{
+                'name': att.name,
+                'type': 'binary',
+                'datas': att.datas,
+                'res_model': event._name,
+                'res_id': event.id
+            } for event in events for att in attachments])
+        return events
+
     def _compute_is_highlighted(self):
-        super(CalendarEvent, self)._compute_is_highlighted()
+        super()._compute_is_highlighted()
         applicant_id = self.env.context.get('active_id')
         if self.env.context.get('active_model') == 'hr.applicant' and applicant_id:
             for event in self:
                 if event.applicant_id.id == applicant_id:
                     event.is_highlighted = True
 
-    applicant_id = fields.Many2one('hr.applicant', string="Applicant", index='btree_not_null', ondelete='set null')
+    @api.depends('applicant_id')
+    def _compute_candidate_id(self):
+        for event in self:
+            if not event.applicant_id:
+                continue
+            event.candidate_id = event.applicant_id.candidate_id

--- a/addons/hr_recruitment/models/digest.py
+++ b/addons/hr_recruitment/models/digest.py
@@ -21,6 +21,6 @@ class Digest(models.Model):
         )
 
     def _compute_kpis_actions(self, company, user):
-        res = super(Digest, self)._compute_kpis_actions(company, user)
-        res['kpi_hr_recruitment_new_colleagues'] = 'hr.open_view_employee_list_my&menu_id=%s' % self.env.ref('hr.menu_hr_root').id
+        res = super()._compute_kpis_actions(company, user)
+        res['kpi_hr_recruitment_new_colleagues'] = f"hr.open_view_employee_list_my&menu_id={self.env.ref('hr.menu_hr_root').id}"
         return res

--- a/addons/hr_recruitment/models/hr_applicant.py
+++ b/addons/hr_recruitment/models/hr_applicant.py
@@ -25,24 +25,38 @@ class Applicant(models.Model):
     _order = "priority desc, id desc"
     _inherit = ['mail.thread.cc',
                'mail.thread.main.attachment',
-               'mail.thread.blacklist',
-               'mail.thread.phone',
                'mail.activity.mixin',
                'utm.mixin',
                'mail.tracking.duration.mixin',
     ]
+    _rec_name = "partner_name"
     _mailing_enabled = True
     _primary_email = 'email_from'
     _track_duration_field = 'stage_id'
 
-    name = fields.Char("Subject / Application", required=True, help="Email subject for applications sent via email", index='trigram')
     active = fields.Boolean("Active", default=True, help="If the active field is set to false, it will allow you to hide the case without removing it.", index=True)
     description = fields.Html("Description")
-    email_from = fields.Char("Email", size=128, compute='_compute_partner_phone_email',
-        inverse='_inverse_partner_email', store=True, index='trigram')
-    email_normalized = fields.Char(index='trigram')  # inherited via mail.thread.blacklist
+
+    candidate_id = fields.Many2one('hr.candidate', required=True, index=True)
+    partner_id = fields.Many2one(related="candidate_id.partner_id")
+    partner_name = fields.Char(related="candidate_id.partner_name", inverse="_inverse_name")
+    email_from = fields.Char(related="candidate_id.email_from", readonly=False)
+    email_normalized = fields.Char(related="candidate_id.email_normalized")
+    partner_phone = fields.Char(related="candidate_id.partner_phone", readonly=False)
+    partner_phone_sanitized = fields.Char(related="candidate_id.partner_phone_sanitized")
+    partner_mobile = fields.Char(related="candidate_id.partner_mobile", readonly=False)
+    partner_mobile_sanitized = fields.Char(related="candidate_id.partner_mobile_sanitized")
+    linkedin_profile = fields.Char(related="candidate_id.linkedin_profile", readonly=False)
+    type_id = fields.Many2one(related="candidate_id.type_id", readonly=False)
+    availability = fields.Date(related="candidate_id.availability", readonly=False)
+    color = fields.Integer(related="candidate_id.color")
+    employee_id = fields.Many2one(related="candidate_id.employee_id", readonly=False)
+    emp_is_active = fields.Boolean(related="candidate_id.emp_is_active")
+    employee_name = fields.Char(related="candidate_id.employee_name")
+
+    similar_candidates_count = fields.Integer(related="candidate_id.similar_candidates_count")
+
     probability = fields.Float("Probability")
-    partner_id = fields.Many2one('res.partner', "Contact", copy=False, index='btree_not_null')
     create_date = fields.Datetime("Applied on", readonly=True)
     stage_id = fields.Many2one('hr.recruitment.stage', 'Stage', ondelete='restrict', tracking=True,
                                compute='_compute_stage', store=True, readonly=False,
@@ -51,7 +65,7 @@ class Applicant(models.Model):
                                group_expand='_read_group_stage_ids')
     last_stage_id = fields.Many2one('hr.recruitment.stage', "Last Stage",
                                     help="Stage of the applicant before being in the current stage. Used for lost cases analysis.")
-    categ_ids = fields.Many2many('hr.applicant.category', string="Tags")
+    categ_ids = fields.Many2many('hr.applicant.category', string="Tags", compute='_compute_categ_ids', store=True, readonly=False)
     company_id = fields.Many2one('res.company', "Company", compute='_compute_company', store=True, readonly=False, tracking=True)
     user_id = fields.Many2one(
         'res.users', "Recruiter", compute='_compute_user', domain="[('share', '=', False), ('company_ids', 'in', company_id)]",
@@ -65,27 +79,14 @@ class Applicant(models.Model):
     salary_expected_extra = fields.Char("Expected Salary Extra", help="Salary Expected by Applicant, extra advantages", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
     salary_proposed = fields.Float("Proposed Salary", aggregator="avg", help="Salary Proposed by the Organisation", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
     salary_expected = fields.Float("Expected Salary", aggregator="avg", help="Salary Expected by Applicant", tracking=True, groups="hr_recruitment.group_hr_recruitment_user")
-    availability = fields.Date("Availability", help="The date at which the applicant will be available to start working", tracking=True)
-    partner_name = fields.Char("Applicant's Name")
-    partner_phone = fields.Char("Phone", size=32, compute='_compute_partner_phone_email',
-        store=True, readonly=False, index='btree_not_null', inverse='_inverse_partner_email')
-    partner_phone_sanitized = fields.Char(string='Sanitized Phone Number', compute='_compute_partner_phone_sanitized', store=True, index='btree_not_null')
-    partner_mobile = fields.Char("Mobile", size=32, compute='_compute_partner_phone_email',
-        store=True, readonly=False, index='btree_not_null', inverse='_inverse_partner_email')
-    partner_mobile_sanitized = fields.Char(string='Sanitized Mobile Number', compute='_compute_partner_mobile_sanitized', store=True, index='btree_not_null')
-    type_id = fields.Many2one('hr.recruitment.degree', "Degree")
     department_id = fields.Many2one(
         'hr.department', "Department", compute='_compute_department', store=True, readonly=False,
         domain="['|', ('company_id', '=', False), ('company_id', '=', company_id)]", tracking=True)
     day_open = fields.Float(compute='_compute_day', string="Days to Open", compute_sudo=True)
     day_close = fields.Float(compute='_compute_day', string="Days to Close", compute_sudo=True)
     delay_close = fields.Float(compute="_compute_delay", string='Delay to Close', readonly=True, aggregator="avg", help="Number of days to close", store=True)
-    color = fields.Integer("Color Index", default=0)
-    emp_id = fields.Many2one('hr.employee', string="Employee", help="Employee linked to the applicant.", copy=False)
-    emp_is_active = fields.Boolean(string="Employee Active", related='emp_id.active')
     user_email = fields.Char(related='user_id.email', string="User Email", readonly=True)
     attachment_number = fields.Integer(compute='_get_attachment_number', string="Number of Attachments")
-    employee_name = fields.Char(related='emp_id.name', string="Employee Name", readonly=False, tracking=False)
     attachment_ids = fields.One2many('ir.attachment', 'res_id', domain=[('res_model', '=', 'hr.applicant')], string='Attachments')
     kanban_state = fields.Selection([
         ('normal', 'Grey'),
@@ -95,7 +96,6 @@ class Applicant(models.Model):
     legend_blocked = fields.Char(related='stage_id.legend_blocked', string='Kanban Blocked')
     legend_done = fields.Char(related='stage_id.legend_done', string='Kanban Valid')
     legend_normal = fields.Char(related='stage_id.legend_normal', string='Kanban Ongoing')
-    application_count = fields.Integer(compute='_compute_application_count', help='Applications with the same email or phone or mobile')
     refuse_reason_id = fields.Many2one('hr.applicant.refuse.reason', string='Refuse Reason', tracking=True)
     meeting_ids = fields.One2many('calendar.event', 'applicant_id', 'Meetings')
     meeting_display_text = fields.Char(compute='_compute_meeting_display')
@@ -107,13 +107,13 @@ class Applicant(models.Model):
     interviewer_ids = fields.Many2many('res.users', 'hr_applicant_res_users_interviewers_rel',
         string='Interviewers', index=True, tracking=True,
         domain="[('share', '=', False), ('company_ids', 'in', company_id)]")
-    linkedin_profile = fields.Char('LinkedIn Profile')
     application_status = fields.Selection([
         ('ongoing', 'Ongoing'),
         ('hired', 'Hired'),
         ('refused', 'Refused'),
         ('archived', 'Archived'),
     ], compute="_compute_application_status", search="_search_application_status")
+    other_applications_count = fields.Integer(compute='_compute_other_applications_count', compute_sudo=True)
     applicant_properties = fields.Properties('Properties', definition='job_id.applicant_properties_definition', copy=True)
     refuse_date = fields.Datetime('Refuse Date')
 
@@ -124,16 +124,18 @@ class Applicant(models.Model):
             ON hr_applicant(job_id, stage_id)
             WHERE active IS TRUE
         """)
-        self.env.cr.execute("""
-            CREATE INDEX IF NOT EXISTS hr_applicant_email_partner_phone_mobile
-            ON hr_applicant(email_normalized, partner_mobile_sanitized, partner_phone_sanitized);
-        """)
 
-    @api.onchange('job_id')
-    def _onchange_job_id(self):
+    def _inverse_name(self):
         for applicant in self:
-            if applicant.job_id.name:
-                applicant.name = applicant.job_id.name
+            if applicant.partner_name and not applicant.candidate_id:
+                applicant.candidate_id = self.env['hr.candidate'].create({'partner_name': applicant.partner_name})
+            else:
+                applicant.candidate_id.partner_name = applicant.partner_name
+
+    @api.depends('candidate_id')
+    def _compute_other_applications_count(self):
+        for applicant in self:
+            applicant.other_applications_count = max(len(applicant.candidate_id.applicant_ids) - 1, 0)
 
     @api.depends('date_open', 'date_closed')
     def _compute_day(self):
@@ -159,62 +161,6 @@ class Applicant(models.Model):
             else:
                 applicant.delay_close = False
 
-    @api.depends('email_from', 'partner_mobile_sanitized', 'partner_phone_sanitized')
-    def _compute_application_count(self):
-        """
-            The field application_count is only used on the form view.
-            Thus, using ORM rather then querying, should not make much
-            difference in terms of performance, while being more readable and secure.
-        """
-        if not any(self._ids):
-            for applicant in self:
-                domain = applicant._get_similar_applicants_domain()
-                if domain:
-                    applicant.application_count = max(0, self.env["hr.applicant"].with_context(active_test=False).search_count(domain) - 1)
-                else:
-                    applicant.application_count = 0
-            return
-        self.flush_recordset(['email_normalized', 'partner_phone_sanitized', 'partner_mobile_sanitized'])
-        self.env.cr.execute("""
-            SELECT
-                id,
-                (
-                    SELECT COUNT(*)
-                    FROM hr_applicant AS sub
-                    WHERE a.id != sub.id
-                     AND ((a.email_normalized <> '' AND sub.email_normalized = a.email_normalized)
-                       OR (a.partner_mobile_sanitized <> '' AND a.partner_mobile_sanitized = sub.partner_mobile_sanitized)
-                       OR (a.partner_mobile_sanitized <> '' AND a.partner_mobile_sanitized = sub.partner_phone_sanitized)
-                       OR (a.partner_phone_sanitized <> '' AND a.partner_phone_sanitized = sub.partner_mobile_sanitized)
-                       OR (a.partner_phone_sanitized <> '' AND a.partner_phone_sanitized = sub.partner_phone_sanitized))
-                ) AS similar_applicants
-            FROM hr_applicant AS a
-            WHERE id IN %(ids)s
-        """, {'ids': tuple(self._origin.ids)})
-        query_results = self.env.cr.dictfetchall()
-        mapped_data = {result['id']: result['similar_applicants'] for result in query_results}
-        for applicant in self:
-            applicant.application_count = mapped_data.get(applicant.id, 0)
-
-    def _get_similar_applicants_domain(self):
-        """
-            This method returns a domain for the applicants whitch match with the
-            current applicant according to email_from, partner_phone or partner_mobile.
-            Thus, search on the domain will return the current applicant as well if any of
-            the following fields are filled.
-        """
-        self.ensure_one()
-        if not self:
-            return None
-        domain = [('id', 'in', self.ids)]
-        if self.email_normalized:
-            domain = expression.OR([domain, [('email_normalized', '=', self.email_normalized)]])
-        if self.partner_phone_sanitized:
-            domain = expression.OR([domain, ['|', ('partner_phone_sanitized', '=', self.partner_phone_sanitized), ('partner_mobile_sanitized', '=', self.partner_phone_sanitized)]])
-        if self.partner_mobile_sanitized:
-            domain = expression.OR([domain, ['|', ('partner_mobile_sanitized', '=', self.partner_mobile_sanitized), ('partner_phone_sanitized', '=', self.partner_mobile_sanitized)]])
-        return domain
-
     @api.depends_context('lang')
     @api.depends('meeting_ids', 'meeting_ids.start')
     def _compute_meeting_display(self):
@@ -239,6 +185,11 @@ class Applicant(models.Model):
             else:
                 applicant.meeting_display_text = _('Last Meeting')
 
+    @api.depends('candidate_id')
+    def _compute_categ_ids(self):
+        for applicant in self:
+            applicant.categ_ids = applicant.candidate_id.categ_ids
+
     @api.depends('refuse_reason_id', 'date_closed')
     def _compute_application_status(self):
         for applicant in self:
@@ -270,11 +221,11 @@ class Applicant(models.Model):
             if status == 'refused':
                 domain = [('refuse_reason_id', '!=', None)]
             elif status == 'hired':
-                domain = [('date_closed', '=', True)]
+                domain = [('date_closed', '!=', False)]
             elif status == 'archived' or status is False:
                 domain = [('active', '=', False)]
             elif status == 'ongoing':
-                domain = [('active', '=', True), ('date_closed', '=', False)]
+                domain = ['&', ('active', '=', True), ('date_closed', '=', False)]
 
         # Invert the domain for '!=' and 'not in' operators
         if operator in expression.NEGATIVE_TERM_OPERATORS:
@@ -338,46 +289,6 @@ class Applicant(models.Model):
         for applicant in self:
             applicant.user_id = applicant.job_id.user_id.id
 
-    @api.depends('partner_id')
-    def _compute_partner_phone_email(self):
-        for applicant in self:
-            if not applicant.partner_id:
-                continue
-            applicant.email_from = applicant.partner_id.email
-            if not applicant.partner_phone:
-                applicant.partner_phone = applicant.partner_id.phone
-            if not applicant.partner_mobile:
-                applicant.partner_mobile = applicant.partner_id.mobile
-
-    def _inverse_partner_email(self):
-        for applicant in self:
-            if not applicant.email_from:
-                continue
-            if not applicant.partner_id:
-                if not applicant.partner_name:
-                    raise UserError(_('You must define a Contact Name for this applicant.'))
-                applicant.partner_id = self.env['res.partner'].with_context(default_lang=self.env.lang).find_or_create(applicant.email_from)
-            if applicant.partner_name and not applicant.partner_id.name:
-                applicant.partner_id.name = applicant.partner_name
-            if tools.email_normalize(applicant.email_from) != tools.email_normalize(applicant.partner_id.email):
-                # change email on a partner will trigger other heavy code, so avoid to change the email when
-                # it is the same. E.g. "email@example.com" vs "My Email" <email@example.com>""
-                applicant.partner_id.email = applicant.email_from
-            if applicant.partner_mobile:
-                applicant.partner_id.mobile = applicant.partner_mobile
-            if applicant.partner_phone:
-                applicant.partner_id.phone = applicant.partner_phone
-
-    @api.depends('partner_phone')
-    def _compute_partner_phone_sanitized(self):
-        for applicant in self:
-            applicant.partner_phone_sanitized = applicant._phone_format(fname='partner_phone') or applicant.partner_phone
-
-    @api.depends('partner_mobile')
-    def _compute_partner_mobile_sanitized(self):
-        for applicant in self:
-            applicant.partner_mobile_sanitized = applicant._phone_format(fname='partner_mobile') or applicant.partner_mobile
-
     def _phone_get_number_fields(self):
         """ This method returns the fields to use to find the number to use to
         send an SMS on a record. """
@@ -391,10 +302,6 @@ class Applicant(models.Model):
             if not applicant.stage_id.hired_stage:
                 applicant.date_closed = False
 
-    def _check_interviewer_access(self):
-        if self.env.user.has_group('hr_recruitment.group_hr_recruitment_interviewer') and not self.env.user.has_group('hr_recruitment.group_hr_recruitment_user'):
-            raise AccessError(_('You are not allowed to perform this action.'))
-
     @api.model_create_multi
     def create(self, vals_list):
         for vals in vals_list:
@@ -413,10 +320,24 @@ class Applicant(models.Model):
                     'applicant_id': applicant.id,
                     'partner_ids': [(6, 0, partners.ids)],
                     'user_id': self.env.uid,
-                    'name': applicant.name,
+                    'name': applicant.partner_name,
                     'start': deadline,
                     'stop': deadline + relativedelta(minutes=30),
                 })
+        # Copy CV from candidate to applicant at record creation
+        attachments_result = self.env['ir.attachment'].read_group([
+            ('res_id', 'in', applicants.candidate_id.ids),
+            ('res_model', '=', "hr.candidate")
+        ], ['ids:array_agg(id)'], groupby=['res_id'])
+        attachments_by_candidate = {e['res_id']: e['ids'] for e in attachments_result}
+        for applicant in applicants:
+            candidate_id = applicant.candidate_id.id
+            if candidate_id not in attachments_by_candidate:
+                continue
+            self.env['ir.attachment'].browse(attachments_by_candidate[candidate_id]).copy({
+                'res_id': applicant.id,
+                'res_model': 'hr.applicant'
+            })
         return applicants
 
     def write(self, vals):
@@ -431,22 +352,19 @@ class Applicant(models.Model):
                 vals['kanban_state'] = 'normal'
             for applicant in self:
                 vals['last_stage_id'] = applicant.stage_id.id
-                res = super(Applicant, self).write(vals)
+                res = super().write(vals)
         else:
-            res = super(Applicant, self).write(vals)
+            res = super().write(vals)
         if 'interviewer_ids' in vals:
             interviewers_to_clean = old_interviewers - self.interviewer_ids
             interviewers_to_clean._remove_recruitment_interviewers()
             self.sudo().interviewer_ids._create_recruitment_interviewers()
-        if vals.get('emp_id'):
-            self._update_employee_from_applicant()
+        if vals.get('date_closed'):
+            for applicant in self:
+                if applicant.job_id.date_to:
+                    applicant.candidate_id.availability = applicant.job_id.date_to + relativedelta(days=1)
         return res
 
-    @api.ondelete(at_uninstall=False)
-    def _unlink_except_linked_employee(self):
-        if self.emp_id:
-            raise UserError(_("The applicant is linked to an employee, to avoid losing information, archive it instead."))
-
     def get_empty_list_help(self, help_message):
         if 'active_id' in self.env.context and self.env.context.get('active_model') == 'hr.job':
             hr_job = self.env['hr.job'].browse(self.env.context['active_id'])
@@ -489,7 +407,7 @@ class Applicant(models.Model):
             view_id = self.env.ref('hr_recruitment.hr_applicant_view_form_interviewer').id
         return super().get_view(view_id, view_type, **options)
 
-    def action_makeMeeting(self):
+    def action_create_meeting(self):
         """ This opens Meeting's calendar view to schedule meeting on current applicant
             @return: Dictionary value for created Meeting view
         """
@@ -517,9 +435,10 @@ class Applicant(models.Model):
         res['context'] = {
             'create': True,
             'default_applicant_id': self.id,
+            'default_candidate_id': self.candidate_id.id,
             'default_partner_ids': partners.ids,
             'default_user_id': self.env.uid,
-            'default_name': self.name,
+            'default_name': self.partner_name,
             'attachment_ids': self.attachment_ids.ids
         }
         return res
@@ -543,33 +462,27 @@ class Applicant(models.Model):
             'domain': [('res_model', '=', 'hr.applicant'), ('res_id', 'in', self.ids), ],
         }
 
-    def action_applications_email(self):
+    def action_open_employee(self):
+        self.ensure_one()
+        return self.candidate_id.action_open_employee()
+
+    def action_open_other_applications(self):
         self.ensure_one()
-        other_applicants = self.env['hr.applicant']
-        domain = self._get_similar_applicants_domain()
-        if domain:
-            other_applicants = self.env['hr.applicant'].with_context(active_test=False).search(domain)
         return {
+            'name': _('Other Applications'),
             'type': 'ir.actions.act_window',
-            'name': _('Job Applications'),
-            'res_model': self._name,
+            'res_model': 'hr.applicant',
             'view_mode': 'tree,kanban,form,pivot,graph,calendar,activity',
-            'domain': [('id', 'in', other_applicants.ids)],
+            'domain': [('id', 'in', (self.candidate_id.applicant_ids - self).ids)],
             'context': {
                 'active_test': False,
                 'search_default_stage': 1,
             },
         }
 
-    def action_open_employee(self):
+    def action_open_similar_candidates(self):
         self.ensure_one()
-        return {
-            'name': _('Employee'),
-            'type': 'ir.actions.act_window',
-            'res_model': 'hr.employee',
-            'view_mode': 'form',
-            'res_id': self.emp_id.id,
-        }
+        return self.candidate_id.action_open_similar_candidates()
 
     def _track_template(self, changes):
         res = super(Applicant, self)._track_template(changes)
@@ -639,18 +552,19 @@ class Applicant(models.Model):
         if custom_values and 'job_id' in custom_values:
             stage = self.env['hr.job'].browse(custom_values['job_id'])._get_first_stage()
         partner_name, email_from_normalized = tools.parse_contact_from_email(msg.get('from'))
+        candidate = self.env['hr.candidate'].create({'partner_name': partner_name or email_from_normalized})
         defaults = {
-            'name': msg.get('subject') or _("No Subject"),
-            'partner_name': partner_name or email_from_normalized,
+            'candidate_id': candidate.id,
         }
         job_platform = self.env['hr.job.platform'].search([('email', '=', email_from_normalized)], limit=1)
         if msg.get('from') and not job_platform:
-            defaults['email_from'] = msg.get('from')
-            defaults['partner_id'] = msg.get('author_id', False)
+            candidate.email_from = msg.get('from')
+            candidate.partner_id = msg.get('author_id', False)
         if msg.get('email_from') and job_platform:
             subject_pattern = re.compile(job_platform.regex or '')
             regex_results = re.findall(subject_pattern, msg.get('subject')) + re.findall(subject_pattern, msg.get('body'))
-            defaults['partner_name'] = regex_results[0] if regex_results else partner_name
+            candidate.partner_name = regex_results[0] if regex_results else partner_name
+            defaults["partner_name"] = candidate.partner_name
             del msg['email_from']
         if msg.get('priority'):
             defaults['priority'] = msg.get('priority')
@@ -659,7 +573,7 @@ class Applicant(models.Model):
         if custom_values:
             defaults.update(custom_values)
         res = super().message_new(msg, custom_values=defaults)
-        res._compute_partner_phone_email()
+        candidate._compute_partner_phone_email()
         return res
 
     def _message_post_after_hook(self, message, msg_vals):
@@ -686,53 +600,17 @@ class Applicant(models.Model):
         return super(Applicant, self)._message_post_after_hook(message, msg_vals)
 
     def create_employee_from_applicant(self):
-        """ Create an employee from applicant """
-        self.ensure_one()
-        self._check_interviewer_access()
-
-        if not self.partner_id:
-            if not self.partner_name:
-                raise UserError(_('Please provide an applicant name.'))
-            self.partner_id = self.env['res.partner'].create({
-                'is_company': False,
-                'name': self.partner_name,
-                'email': self.email_from,
-            })
-
-        action = self.env['ir.actions.act_window']._for_xml_id('hr.open_view_employee_list')
-        employee = self.env['hr.employee'].create(self._get_employee_create_vals())
-        action['res_id'] = employee.id
-        return action
-
-    def _get_employee_create_vals(self):
         self.ensure_one()
-        address_id = self.partner_id.address_get(['contact'])['contact']
-        address_sudo = self.env['res.partner'].sudo().browse(address_id)
-        return {
-            'name': self.partner_name or self.partner_id.display_name,
-            'work_contact_id': self.partner_id.id,
+        action = self.candidate_id.create_employee_from_candidate()
+        employee = self.env['hr.employee'].browse(action['res_id'])
+        employee.write({
             'job_id': self.job_id.id,
             'job_title': self.job_id.name,
-            'private_street': address_sudo.street,
-            'private_street2': address_sudo.street2,
-            'private_city': address_sudo.city,
-            'private_state_id': address_sudo.state_id.id,
-            'private_zip': address_sudo.zip,
-            'private_country_id': address_sudo.country_id.id,
-            'private_phone': address_sudo.phone,
-            'private_email': address_sudo.email,
-            'lang': address_sudo.lang,
             'department_id': self.department_id.id,
-            'address_id': self.company_id.partner_id.id,
             'work_email': self.department_id.company_id.email or self.email_from, # To have a valid email address by default
             'work_phone': self.department_id.company_id.phone,
-            'applicant_id': self.ids,
-            'phone': self.partner_phone or self.partner_mobile
-        }
-
-    def _update_employee_from_applicant(self):
-        # This method is to be overriden
-        return
+        })
+        return action
 
     def archive_applicant(self):
         return {

--- a/addons/hr_recruitment/models/hr_applicant_category.py
+++ b/addons/hr_recruitment/models/hr_applicant_category.py
@@ -1,9 +1,10 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
+from random import randint
+
 from odoo import fields, models
 
-from random import randint
 
 class ApplicantCategory(models.Model):
     _name = "hr.applicant.category"

--- a/None
+++ b/addons/hr_recruitment/models/hr_candidate.py
@@ -0,0 +1,413 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from collections import defaultdict
+
+from odoo.addons.hr_recruitment.models.hr_applicant import AVAILABLE_PRIORITIES
+
+from odoo import api, models, fields, SUPERUSER_ID, tools, _
+from odoo.exceptions import UserError
+from odoo.osv import expression
+
+
+class HrCandidate(models.Model):
+    _name = "hr.candidate"
+    _description = "Candidate"
+    _inherit = ['mail.thread.cc',
+               'mail.thread.main.attachment',
+               'mail.thread.blacklist',
+               'mail.thread.phone',
+               'mail.activity.mixin',
+    ]
+    _order = "priority desc, availability asc, id desc"
+    _mailing_enabled = True
+    _primary_email = 'email_from'
+    _rec_name = 'partner_name'
+
+    active = fields.Boolean("Active", default=True, index=True)
+    company_id = fields.Many2one('res.company', "Company", default=lambda self: self.env.company)
+    applicant_ids = fields.One2many('hr.applicant', 'candidate_id')
+    application_count = fields.Integer(compute="_compute_application_count")
+    partner_id = fields.Many2one('res.partner', "Contact", copy=False, index='btree_not_null')
+    partner_name = fields.Char("Candidates's Name")
+    email_from = fields.Char(
+        string="Email",
+        size=128,
+        compute='_compute_partner_phone_email',
+        inverse='_inverse_partner_email',
+        store=True,
+        index='trigram')
+    email_normalized = fields.Char(index='trigram')  # inherited via mail.thread.blacklist
+    partner_phone = fields.Char(
+        string="Phone",
+        size=32,
+        compute='_compute_partner_phone_email',
+        inverse='_inverse_partner_email',
+        store=True,
+        index='btree_not_null')
+    partner_phone_sanitized = fields.Char(
+        string='Sanitized Phone Number',
+        compute='_compute_partner_phone_sanitized',
+        store=True,
+        index='btree_not_null')
+    partner_mobile = fields.Char(
+        string="Mobile",
+        size=32,
+        compute='_compute_partner_phone_email',
+        inverse='_inverse_partner_email',
+        store=True,
+        index='btree_not_null')
+    partner_mobile_sanitized = fields.Char(
+        string='Sanitized Mobile Number',
+        compute='_compute_partner_mobile_sanitized',
+        store=True,
+        index='btree_not_null')
+    linkedin_profile = fields.Char('LinkedIn Profile')
+    type_id = fields.Many2one('hr.recruitment.degree', "Degree")
+    availability = fields.Date("Availability", help="The date at which the applicant will be available to start working", tracking=True)
+    categ_ids = fields.Many2many('hr.applicant.category', string="Tags")
+    color = fields.Integer("Color Index", default=0)
+    priority = fields.Selection(AVAILABLE_PRIORITIES, string="Evaluation", compute="_compute_priority", store=True)
+    user_id = fields.Many2one(
+        'res.users',
+        string="Candidate Manager",
+        default=lambda self: self.env.user if self.env.user.id != SUPERUSER_ID else False,
+        domain="[('share', '=', False), ('company_ids', 'in', company_id)]",
+        tracking=True)
+    employee_id = fields.Many2one('hr.employee', string="Employee", help="Employee linked to the candidate.", copy=False)
+    emp_is_active = fields.Boolean(string="Employee Active", related='employee_id.active')
+    employee_name = fields.Char(related='employee_id.name', string="Employee Name", readonly=False, tracking=False)
+
+    similar_candidates_count = fields.Integer(
+        compute='_compute_similar_candidates_count',
+        help='Candidates with the same email or phone or mobile')
+    applications_count = fields.Integer(string="# Offers", compute='_compute_applications_count')
+    refused_applications_count = fields.Integer(string="# Refused Offers", compute='_compute_applications_count')
+    accepted_applications_count = fields.Integer(string="# Accepted Offers", compute='_compute_applications_count')
+
+    meeting_ids = fields.One2many('calendar.event', 'candidate_id', 'Meetings')
+    meeting_display_text = fields.Char(compute='_compute_meeting_display')
+    meeting_display_date = fields.Date(compute='_compute_meeting_display')
+    attachment_count = fields.Integer(
+        string="Number of Attachments",
+        compute='_compute_attachment_count')
+    candidate_properties = fields.Properties('Properties', definition='company_id.candidate_properties_definition', copy=True)
+    attachment_ids = fields.One2many('ir.attachment', 'res_id', domain=[('res_model', '=', 'hr.candidate')], string='Attachments')
+
+    def init(self):
+        self.env.cr.execute("""
+            CREATE INDEX IF NOT EXISTS hr_candidate_email_partner_phone_mobile
+            ON hr_candidate(email_normalized, partner_mobile_sanitized, partner_phone_sanitized);
+        """)
+
+    @api.depends('partner_name')
+    def _compute_display_name(self):
+        for candidate in self:
+            candidate.display_name = candidate.partner_name or candidate.partner_id.name
+
+    @api.depends('partner_phone')
+    def _compute_partner_phone_sanitized(self):
+        for candidate in self:
+            candidate.partner_phone_sanitized = candidate._phone_format(fname='partner_phone') or candidate.partner_phone
+
+    @api.depends('partner_mobile')
+    def _compute_partner_mobile_sanitized(self):
+        for candidate in self:
+            candidate.partner_mobile_sanitized = candidate._phone_format(fname='partner_mobile') or candidate.partner_mobile
+
+    @api.depends('partner_id')
+    def _compute_partner_phone_email(self):
+        for candidate in self:
+            if not candidate.partner_id:
+                continue
+            candidate.email_from = candidate.partner_id.email
+            if not candidate.partner_phone:
+                candidate.partner_phone = candidate.partner_id.phone
+            if not candidate.partner_mobile:
+                candidate.partner_mobile = candidate.partner_id.mobile
+
+    def _inverse_partner_email(self):
+        for candidate in self:
+            if not candidate.email_from:
+                continue
+            if not candidate.partner_id:
+                if not candidate.partner_name:
+                    raise UserError(_('You must define a Contact Name for this candidate.'))
+                candidate.partner_id = self.env['res.partner'].with_context(default_lang=self.env.lang).find_or_create(candidate.email_from)
+            if candidate.partner_name and not candidate.partner_id.name:
+                candidate.partner_id.name = candidate.partner_name
+            if tools.email_normalize(candidate.email_from) != tools.email_normalize(candidate.partner_id.email):
+                # change email on a partner will trigger other heavy code, so avoid to change the email when
+                # it is the same. E.g. "email@example.com" vs "My Email" <email@example.com>""
+                candidate.partner_id.email = candidate.email_from
+            if candidate.partner_mobile:
+                candidate.partner_id.mobile = candidate.partner_mobile
+            if candidate.partner_phone:
+                candidate.partner_id.phone = candidate.partner_phone
+
+    @api.depends('email_from', 'partner_mobile_sanitized', 'partner_phone_sanitized')
+    def _compute_similar_candidates_count(self):
+        """
+            The field similar_candidates_count is only used on the form view.
+            Thus, using ORM rather then querying, should not make much
+            difference in terms of performance, while being more readable and secure.
+        """
+        if not any(self._ids):
+            for candidate in self:
+                domain = candidate._get_similar_candidates_domain()
+                if domain:
+                    candidate.similar_candidates_count = max(0, self.env["hr.candidate"].with_context(active_test=False).search_count(domain) - 1)
+                else:
+                    candidate.similar_candidates_count = 0
+            return
+        self.flush_recordset(['email_normalized', 'partner_phone_sanitized', 'partner_mobile_sanitized'])
+        self.env.cr.execute("""
+            SELECT
+                id,
+                (
+                    SELECT COUNT(*)
+                    FROM hr_candidate AS sub
+                    WHERE c.id != sub.id
+                     AND ((coalesce(c.email_normalized, '') <> '' AND sub.email_normalized = c.email_normalized)
+                       OR (coalesce(c.partner_mobile_sanitized, '') <> '' AND c.partner_mobile_sanitized = sub.partner_mobile_sanitized)
+                       OR (coalesce(c.partner_mobile_sanitized, '') <> '' AND c.partner_mobile_sanitized = sub.partner_phone_sanitized)
+                       OR (coalesce(c.partner_phone_sanitized, '') <> '' AND c.partner_phone_sanitized = sub.partner_mobile_sanitized)
+                       OR (coalesce(c.partner_phone_sanitized, '') <> '' AND c.partner_phone_sanitized = sub.partner_phone_sanitized))
+                ) AS similar_candidates
+            FROM hr_candidate AS c
+            WHERE id IN %(ids)s
+        """, {'ids': tuple(self._origin.ids)})
+        query_results = self.env.cr.dictfetchall()
+        mapped_data = {result['id']: result['similar_candidates'] for result in query_results}
+        for candidate in self:
+            candidate.similar_candidates_count = mapped_data.get(candidate.id, 0)
+
+    def _get_similar_candidates_domain(self):
+        """
+            This method returns a domain for the applicants whitch match with the
+            current candidate according to email_from, partner_phone or partner_mobile.
+            Thus, search on the domain will return the current candidate as well if any of
+            the following fields are filled.
+        """
+        self.ensure_one()
+        if not self:
+            return []
+        domain = [('id', 'in', self.ids)]
+        if self.email_normalized:
+            domain = expression.OR([domain, [('email_normalized', '=', self.email_normalized)]])
+        if self.partner_phone_sanitized:
+            domain = expression.OR([domain, ['|', ('partner_phone_sanitized', '=', self.partner_phone_sanitized), ('partner_mobile_sanitized', '=', self.partner_phone_sanitized)]])
+        if self.partner_mobile_sanitized:
+            domain = expression.OR([domain, ['|', ('partner_mobile_sanitized', '=', self.partner_mobile_sanitized), ('partner_phone_sanitized', '=', self.partner_mobile_sanitized)]])
+        return domain
+
+    def _compute_attachment_count(self):
+        read_group_res = self.env['ir.attachment']._read_group(
+            [('res_model', '=', 'hr.candidate'), ('res_id', 'in', self.ids)],
+            ['res_id'], ['__count'])
+        attach_data = dict(read_group_res)
+        for candidate in self:
+            candidate.attachment_count = attach_data.get(candidate.id, 0)
+
+    def _compute_application_count(self):
+        read_group_res = self.env['hr.applicant']._read_group(
+            [('candidate_id', 'in', self.ids)],
+            ['candidate_id'], ['__count'])
+        application_data = dict(read_group_res)
+        for candidate in self:
+            candidate.application_count = application_data.get(candidate, 0)
+
+    @api.depends('applicant_ids.priority')
+    def _compute_priority(self):
+        for candidate in self:
+            if not candidate.applicant_ids:
+                candidate.priority = "0"
+            else:
+                candidate.priority = str(round(sum(int(a.priority) for a in candidate.applicant_ids) / len(candidate.applicant_ids)))
+
+    def _compute_applications_count(self):
+        result = defaultdict(lambda: {"total": 0, "refused": 0, "accepted": 0})
+        for applicant in self.with_context(active_test=False).applicant_ids:
+            result[applicant.candidate_id.id]["total"] += 1
+            if applicant.application_status == "refused":
+                result[applicant.candidate_id.id]["refused"] += 1
+            elif applicant.application_status == "hired":
+                result[applicant.candidate_id.id]["accepted"] += 1
+        for candidate in self:
+            candidate.applications_count = result[candidate.id]['total']
+            candidate.refused_applications_count = result[candidate.id]['refused']
+            candidate.accepted_applications_count = result[candidate.id]['accepted']
+
+    @api.depends_context('lang')
+    @api.depends('meeting_ids', 'meeting_ids.start')
+    def _compute_meeting_display(self):
+        candidate_with_meetings = self.filtered('meeting_ids')
+        (self - candidate_with_meetings).update({
+            'meeting_display_text': _('No Meeting'),
+            'meeting_display_date': ''
+        })
+        today = fields.Date.today()
+        for candidate in candidate_with_meetings:
+            count = len(candidate.meeting_ids)
+            dates = candidate.meeting_ids.mapped('start')
+            min_date, max_date = min(dates).date(), max(dates).date()
+            if min_date >= today:
+                candidate.meeting_display_date = min_date
+            else:
+                candidate.meeting_display_date = max_date
+            if count == 1:
+                candidate.meeting_display_text = _('1 Meeting')
+            elif candidate.meeting_display_date >= today:
+                candidate.meeting_display_text = _('Next Meeting')
+            else:
+                candidate.meeting_display_text = _('Last Meeting')
+
+    def action_open_similar_candidates(self):
+        self.ensure_one()
+        domain = self._get_similar_candidates_domain()
+        similar_candidates = self.env['hr.candidate'].with_context(active_test=False).search(domain)
+        return {
+            'type': 'ir.actions.act_window',
+            'name': _('Similar Candidates'),
+            'res_model': self._name,
+            'view_mode': 'tree,kanban,form,activity',
+            'domain': [('id', 'in', similar_candidates.ids)],
+            'context': {
+                'active_test': False,
+            },
+        }
+
+    def action_open_applications(self):
+        self.ensure_one()
+        return {
+            'name': _('Applications'),
+            'type': 'ir.actions.act_window',
+            'res_model': 'hr.applicant',
+            'view_mode': 'tree,kanban,form,pivot,graph,calendar,activity',
+            'domain': [('id', 'in', self.applicant_ids.ids)],
+            'context': {
+                'active_test': False,
+                'search_default_stage': 1,
+            },
+        }
+
+    def action_open_employee(self):
+        self.ensure_one()
+        return {
+            'name': _('Employee'),
+            'type': 'ir.actions.act_window',
+            'res_model': 'hr.employee',
+            'view_mode': 'form',
+            'res_id': self.employee_id.id,
+        }
+
+    def action_create_meeting(self):
+        """ This opens Meeting's calendar view to schedule meeting on current candidate
+            @return: Dictionary value for created Meeting view
+        """
+        self.ensure_one()
+        if not self.partner_id:
+            if not self.partner_name:
+                raise UserError(_('You must define a Contact Name for this candidate.'))
+            self.partner_id = self.env['res.partner'].create({
+                'is_company': False,
+                'name': self.partner_name,
+                'email': self.email_from,
+            })
+
+        partners = self.partner_id
+        if self.env.user.has_group('hr_recruitment.group_hr_recruitment_interviewer') and not self.env.user.has_group('hr_recruitment.group_hr_recruitment_user'):
+            partners |= self.env.user.partner_id
+        else:
+            partners |= self.user_id.partner_id
+
+        res = self.env['ir.actions.act_window']._for_xml_id('calendar.action_calendar_event')
+        # As we are redirected from the hr.candidate, calendar checks rules on "hr.applicant",
+        # in order to decide whether to allow creation of a meeting.
+        # As interviewer does not have create right on the hr.applicant, in order to allow them
+        # to create a meeting for an applicant, we pass 'create': True to the context.
+        res['context'] = {
+            'create': True,
+            'default_candidate_id': self.id,
+            'default_partner_ids': partners.ids,
+            'default_user_id': self.env.uid,
+            'default_name': self.partner_name,
+            'attachment_ids': self.attachment_ids.ids
+        }
+        return res
+
+    def write(self, vals):
+        res = super().write(vals)
+        if vals.get('employee_id'):
+            self._update_employee_from_candidate()
+        return res
+
+    def _update_employee_from_candidate(self):
+        # This method is to be overriden
+        return
+
+    @api.ondelete(at_uninstall=False)
+    def _unlink_except_linked_employee(self):
+        if self.employee_id:
+            raise UserError(_("The candidate is linked to an employee, to avoid losing information, archive it instead."))
+
+    def create_employee_from_candidate(self):
+        self.ensure_one()
+        self._check_interviewer_access()
+
+        if not self.partner_id:
+            if not self.partner_name:
+                raise UserError(_('Please provide an candidate name.'))
+            self.partner_id = self.env['res.partner'].create({
+                'is_company': False,
+                'name': self.partner_name,
+                'email': self.email_from,
+            })
+
+        action = self.env['ir.actions.act_window']._for_xml_id('hr.open_view_employee_list')
+        employee = self.env['hr.employee'].create(self._get_employee_create_vals())
+        action['res_id'] = employee.id
+        return action
+
+    def _get_employee_create_vals(self):
+        self.ensure_one()
+        address_id = self.partner_id.address_get(['contact'])['contact']
+        address_sudo = self.env['res.partner'].sudo().browse(address_id)
+        return {
+            'name': self.partner_name or self.partner_id.display_name,
+            'work_contact_id': self.partner_id.id,
+            'private_street': address_sudo.street,
+            'private_street2': address_sudo.street2,
+            'private_city': address_sudo.city,
+            'private_state_id': address_sudo.state_id.id,
+            'private_zip': address_sudo.zip,
+            'private_country_id': address_sudo.country_id.id,
+            'private_phone': address_sudo.phone,
+            'private_email': address_sudo.email,
+            'lang': address_sudo.lang,
+            'address_id': self.company_id.partner_id.id,
+            'candidate_id': self.ids,
+            'phone': self.partner_phone or self.partner_mobile
+        }
+
+    def _check_interviewer_access(self):
+        if self.env.user.has_group('hr_recruitment.group_hr_recruitment_interviewer') and not self.env.user.has_group('hr_recruitment.group_hr_recruitment_user'):
+            raise UserError(_('You are not allowed to perform this action.'))
+
+    def action_open_attachments(self):
+        return {
+            'type': 'ir.actions.act_window',
+            'res_model': 'ir.attachment',
+            'name': _('Documents'),
+            'context': {
+                'default_res_model': 'hr.candidate',
+                'default_res_id': self.ids[0],
+                'show_partner_name': 1,
+            },
+            'view_mode': 'tree,form',
+            'views': [
+                (self.env.ref('hr_recruitment.ir_attachment_hr_recruitment_list_view').id, 'tree'),
+                (False, 'form'),
+            ],
+            'search_view_id': self.env.ref('hr_recruitment.ir_attachment_view_search_inherit_hr_recruitment').ids,
+            'domain': [('res_model', '=', 'hr.candidate'), ('res_id', 'in', self.ids)],
+        }

--- a/addons/hr_recruitment/models/hr_employee.py
+++ b/addons/hr_recruitment/models/hr_employee.py
@@ -1,28 +1,28 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, fields, models
-from datetime import timedelta
 
 
 class HrEmployee(models.Model):
     _inherit = "hr.employee"
 
-    applicant_id = fields.One2many('hr.applicant', 'emp_id', 'Applicant', groups="hr.group_hr_user")
+    # YTI Rename into candidate_ids
+    candidate_id = fields.One2many('hr.candidate', 'employee_id', 'Candidate', groups="hr.group_hr_user")
 
     def _get_partner_count_depends(self):
-        return super()._get_partner_count_depends() + ['applicant_id']
+        return super()._get_partner_count_depends() + ['candidate_id']
 
     def _get_related_partners(self):
         partners = super()._get_related_partners()
-        return partners | self.sudo().applicant_id.partner_id
+        return partners | self.sudo().candidate_id.partner_id
 
     @api.model_create_multi
     def create(self, vals_list):
         employees = super().create(vals_list)
         for employee in employees:
-            if employee.applicant_id:
-                employee.applicant_id._message_log_with_view(
-                    'hr_recruitment.applicant_hired_template',
-                    render_values={'applicant': employee.applicant_id}
+            if employee.candidate_id:
+                employee.candidate_id._message_log_with_view(
+                    'hr_recruitment.candidate_hired_template',
+                    render_values={'candidate': employee.candidate_id}
                 )
         return employees

--- a/addons/hr_recruitment/models/hr_job.py
+++ b/addons/hr_recruitment/models/hr_job.py
@@ -55,6 +55,9 @@ class Job(models.Model):
     favorite_user_ids = fields.Many2many('res.users', 'job_favorite_user_rel', 'job_id', 'user_id', default=_get_default_favorite_user_ids)
     interviewer_ids = fields.Many2many('res.users', string='Interviewers', domain="[('share', '=', False), ('company_ids', 'in', company_id)]", help="The Interviewers set on the job position can see all Applicants in it. They have access to the information, the attachments, the meeting management and they can refuse him. You don't need to have Recruitment rights to be set as an interviewer.")
     extended_interviewer_ids = fields.Many2many('res.users', 'hr_job_extended_interviewer_res_users', compute='_compute_extended_interviewer_ids', store=True)
+    industry_id = fields.Many2one('res.partner.industry', 'Industry')
+    date_from = fields.Date(help="Is set, update candidates availability once hired for that specific mission.")
+    date_to = fields.Date()
 
     activities_overdue = fields.Integer(compute='_compute_activities')
     activities_today = fields.Integer(compute='_compute_activities')
@@ -139,7 +142,7 @@ class Job(models.Model):
         unfavorited_jobs.write({'favorite_user_ids': [(3, self.env.uid)]})
 
     def _compute_document_ids(self):
-        applicants = self.mapped('application_ids').filtered(lambda self: not self.emp_id)
+        applicants = self.mapped('application_ids').filtered(lambda self: not self.employee_id)
         app_to_job = dict((applicant.id, applicant.job_id.id) for applicant in applicants)
         attachments = self.env['ir.attachment'].search([
             '|',
@@ -231,7 +234,7 @@ class Job(models.Model):
             job.old_application_count = job.application_count - job.new_application_count
 
     def _alias_get_creation_values(self):
-        values = super(Job, self)._alias_get_creation_values()
+        values = super()._alias_get_creation_values()
         values['alias_model_id'] = self.env['ir.model']._get('hr.applicant').id
         if self.id:
             values['alias_defaults'] = defaults = ast.literal_eval(self.alias_defaults or "{}")

--- a/addons/hr_recruitment/models/hr_recruitment_degree.py
+++ b/addons/hr_recruitment/models/hr_recruitment_degree.py
@@ -7,9 +7,10 @@ from odoo import fields, models
 class RecruitmentDegree(models.Model):
     _name = "hr.recruitment.degree"
     _description = "Applicant Degree"
-    _sql_constraints = [
-        ('name_uniq', 'unique (name)', 'The name of the Degree of Recruitment must be unique!')
-    ]
 
     name = fields.Char("Degree Name", required=True, translate=True)
     sequence = fields.Integer("Sequence", default=1)
+
+    _sql_constraints = [
+        ('name_uniq', 'unique (name)', 'The name of the Degree of Recruitment must be unique!')
+    ]

--- a/None
+++ b/addons/hr_recruitment/models/res_company.py
@@ -0,0 +1,9 @@
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    candidate_properties_definition = fields.PropertiesDefinition('Candidate Properties')

--- a/addons/hr_recruitment/models/res_config_settings.py
+++ b/addons/hr_recruitment/models/res_config_settings.py
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import fields, models
